_rules/table-headers-attribute-refer-to-data-cells-a25f45.md

```diff
@@ -0,0 +1,287 @@
+---
+id: a25f45
+name: Headers attribute of td element refers to cells in the same table element
+rule_type: atomic
+description: |
+  This rule checks that the headers attribute of `td` elements refer to cells in the same `table` element that have a semantic role of `columnheader` or `rowheader`.
+accessibility_requirements:
+  wcag20:1.3.1: # Info and Relationships (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+authors:
+  - Jey Nandakumar
+  - Audrey Maniez
+---
+
+## Applicability
+
+This rule applies to any [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).
+
+## Expectation
+
+The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:
+
+1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and
```

Comment:
I don't really understand why this is required. I know the spec says that you should only reference other headers with the `headers` attribute, but are there any accessibility issues that come from referencing a `td` element with the `headers` attribute? I don't see what they would be, unless maybe assistive tech don't announce `td`s if they're used in the `headers` attribute, but I don't think that's the case.

Reply:
Note to self:
- Get data of AT testing.

Reply:
@WilcoFiers 

I did some testing & other `td` referred via `headers` attribute is not consistently announced without a role of `columnheder` or `rowheader`. These tests were done with voiceover.

Note voiceover seems to be very inconsistent even when `th` is referred in `headers` attribute. It also seems like headers are only rightly announced when wrapped with `<thead>` in the case of voiceover.

Here is also an article on elaborate testing outcomes:
- [VoiceOver and Tables](http://accessibleculture.org/articles/2010/10/voiceover-and-tables-with-an-empty-first-header-cell/)

I did not test all of the test cases listed in the above example.

Note though in terms of automated implementation of this rule, I do think it does not matter whether a header attribute refers to `th` or `td`.

Reply:
Did more testing referencing `td` with `headers` attribute, does not get announced in VoiceOver :/. Overall headers attribute is quite inconsistent with AT's.